### PR TITLE
review refactor: all non-leaf interfaces of the metamodel should be visited by CtInheritanceScanner

### LINKED ISSUE #1430: bug: CtBodyHolder, CTRHSReceiver and CtShadowable are not CtInheritanceScan-ABLE
See pb with CtBodyHolder, CTRHSReceiver and CtShadowable

one can check this contract in a new architectural test

### PR DESCRIPTION
**Pull request by surli**
Fix #1430

### CODE REVIEW

**Comment by surli** [comment]
@monperrus in #1430 you talked about 
> every abstract non-leaf CtElement

but how do you define a leaf CtElement? If I'm comparing against the package location, there is some interfaces in `spoon.reflect.code` which are non leaf. It's not perfectly clear for me which interface should be scannable and which one should not be. I currently get the following list of missing method in `CtInheritanceScanner`:
<details>
scanCtAnnotationFieldAccess
scanCtArrayRead
scanCtArrayWrite
scanCtAssert
scanCtAssignment
scanCtBinaryOperator
scanCtBlock
scanCtBreak
scanCtCase
scanCtCatch
scanCtCatchVariable
scanCtCodeSnippetExpression
scanCtCodeSnippetStatement
scanCtComment
scanCtConditional
scanCtConstructorCall
scanCtContinue
scanCtDo
scanCtExecutableReferenceExpression
scanCtFieldRead
scanCtFieldWrite
scanCtFor
scanCtForEach
scanCtIf
scanCtInvocation
scanCtJavaDoc
scanCtLambda
scanCtLiteral
scanCtLocalVariable
scanCtNewArray
scanCtNewClass
scanCtOperatorAssignment
scanCtReturn
scanCtStatementList
scanCtSuperAccess
scanCtSwitch
scanCtSynchronized
scanCtThisAccess
scanCtThrow
scanCtTry
scanCtTryWithResource
scanCtTypeAccess
scanCtUnaryOperator
scanCtVariableRead
scanCtVariableWrite
scanCtWhile
scanCtAnnotation
scanCtAnnotationMethod
scanCtAnnotationType
scanCtAnonymousExecutable
scanCtClass
scanCtConstructor
scanCtEnum
scanCtEnumValue
scanCtField
scanCtInterface
scanCtMethod
scanCtPackage
scanCtParameter
scanCtTypeParameter
scanCtArrayTypeReference
scanCtCatchVariableReference
scanCtExecutableReference
scanCtFieldReference
scanCtIntersectionTypeReference
scanCtLocalVariableReference
scanCtPackageReference
scanCtParameterReference
scanCtTypeParameterReference
scanCtTypeReference
scanCtUnboundVariableReference
scanCtWildcardReference
</details>

**Comment by pvojtechovsky** [comment]
> every abstract non-leaf CtElement

This might work:
```java
new SpoonMetaModel(getFactory()).getMMTypes().forEach(mmType->{
  if(mmType.getKind()==ABSTRACT && mmType.getModelInteface()!=null)  {
    CtInterface abstractIface = mmType.getModelInteface();
    //abstractIface is called with each abstract interface of spoon model
  }
});
```

File: src/test/java/spoon/test/architecture/SpoonArchitectureEnforcerTest.java [hunk 1]
```diff
@@ -206,4 +212,32 @@ public boolean matches(CtClass element) {
 			assertTrue(klass.getElements(new TypeFilter<>(CtConstructor.class)).stream().allMatch(x -> x.hasModifier(ModifierKind.PRIVATE)));
 		}
 	}
+
+	@Test
+	public void testInterfacesAreCtScannable() {
+		// contract: all interface inherited from CtElement should be visited in CtScanner
```
Comment (monperrus) [inline src/test/java/spoon/test/architecture/SpoonArchitectureEnforcerTest.java]: the contract of this bug should be:
all non-leaf interfaces of the metamodel should be visited by CtInheritance scanner

the interfaces do not have to inherit from ctelement

File: src/test/java/spoon/test/architecture/SpoonArchitectureEnforcerTest.java [hunk 2]
```diff
@@ -217,26 +219,26 @@ public boolean matches(CtClass element) {
...
 
 		List<String> missingMethods = new ArrayList<>();
-		for (CtType type : allTypes) {
-			if (type.isSubtypeOf(ctElement.getReference())) {
-				String methodName = "scan"+type.getSimpleName();
+
+		new SpoonMetaModel(interfaces.getFactory()).getMMTypes().forEach(mmType->{
+			if(mmType.getKind()==ABSTRACT && mmType.getModelInterface() != null)  {
+				CtInterface abstractIface = mmType.getModelInterface();
+				String methodName = "scan"+abstractIface.getSimpleName();
```
Comment (pvojtechovsky) [inline src/test/java/spoon/test/architecture/SpoonArchitectureEnforcerTest.java]: Thes 2 lines above may be replaced by shorter
```
String methodName = "scan"+mmType.getName();
```

**Comment by monperrus** [comment]
@pvojtechovsky would you merge this one?

File: src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java [hunk 3]
```diff
@@ -567,7 +567,7 @@ public void visitCtCatch(CtCatch catchBlock) {
 		elementPrinterHelper.writeComment(catchBlock, CommentOffset.BEFORE);
 		printer.writeSpace().writeKeyword("catch").writeSpace().writeSeparator("(");
 		CtCatchVariable<? extends Throwable> parameter = catchBlock.getParameter();
-		if (parameter.getMultiTypes().size() > 1) {
+		if (parameter != null && parameter.getMultiTypes() != null && parameter.getMultiTypes().size() > 1) {
```
Comment (pvojtechovsky) [inline src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java]: parameter.getMultiTypes() is never null. I suggest to remove this check. 
WDYT?

File: src/test/java/spoon/test/architecture/SpoonArchitectureEnforcerTest.java [hunk 4]
```diff
@@ -212,6 +216,34 @@ public boolean matches(CtClass element) {
...
 	@Test
+	public void testInterfacesAreCtScannable() {
+		// contract: all non-leaf interfaces of the metamodel should be visited by CtInheritanceScanner
+		Launcher interfaces = new Launcher();
+		interfaces.addInputResource("src/main/java/spoon/support");
+		interfaces.addInputResource("src/main/java/spoon/reflect/declaration");
+		interfaces.addInputResource("src/main/java/spoon/reflect/code");
+		interfaces.addInputResource("src/main/java/spoon/reflect/reference");
+		interfaces.addInputResource("src/main/java/spoon/reflect/visitor/CtScanner.java");
+		interfaces.buildModel();
```
Comment (pvojtechovsky) [inline src/test/java/spoon/test/architecture/SpoonArchitectureEnforcerTest.java]: SpoonMetaModel needs these packages too
				"spoon/support/reflect/declaration",
				"spoon/support/reflect/code",
				"spoon/support/reflect/reference"
Otherwise meta model is not complete.

File: src/test/java/spoon/test/architecture/SpoonArchitectureEnforcerTest.java [hunk 5]
```diff
@@ -212,6 +216,34 @@ public boolean matches(CtClass element) {
 	}
 
 	@Test
+	public void testInterfacesAreCtScannable() {
+		// contract: all non-leaf interfaces of the metamodel should be visited by CtInheritanceScanner
```
Comment (pvojtechovsky) [inline src/test/java/spoon/test/architecture/SpoonArchitectureEnforcerTest.java]: This test checks that there is a scan method. It is good test. But there is not checked that visitXxx method calls all scan methods for each interface implemented by that non abstract class. I suggest to add this test too. May be as part of another PR?
WDYT?

**Comment by pvojtechovsky** [comment]
Thank You Simon for that work! 
I do not persist on fixing of my suggestions. They are optional. Just let me know if you are done. Then this PR is OK for me.

File: src/test/java/spoon/test/architecture/SpoonArchitectureEnforcerTest.java (hunk 5 above)
Comment (surli) [reply src/test/java/spoon/test/architecture/SpoonArchitectureEnforcerTest.java]: I thought it was the goal of https://github.com/INRIA/spoon/blob/65c1b2974af8103b3b44f6f6306a310e6552e868/src/test/java/spoon/reflect/visitor/CtInheritanceScannerTest.java#L118

**Comment by pvojtechovsky** [comment]
> there is not checked that visitXxx method calls all scan methods
> https://github.com/INRIA/spoon/blob/65c1b2974af8103b3b44f6f6306a310e6552e868/src/test/java/spoon/reflect/visitor/CtInheritanceScannerTest.java#L118

You are right, so there is already a test for that.
